### PR TITLE
BUG, BLD: Generate the main dispatcher config header into the build dir

### LINKED ISSUE #18092: CCompilerOpt.new_ccompiler_opt calls generate_dispatch_header on NumPy install directory
Trying to build scipy-1.5.4 or 1.6.0rc2 with numpy-1.20.0rc2 installed as a system library (via dpkg) fails on a distutils call to `new_ccompiler_opt()`.

### Reproducing code example:
Running the scipy build with
`ATLAS=None python3.8 setup.py build --fcompiler gnu95`
fails with either

### Error message:
```
CCompilerOpt.new_ccompiler_opt[2529] : dispatch header dir '/opt/lib/python3.8/site-packages/numpy/core/src/common' isn't (!) exist, creating it
error: [Errno 13] Permission denied: '/opt/lib/python3.8/site-packages/numpy/core/src'
CCompilerOpt._cache_write[796] : write cache to path -> /private/var/tmp/fink.build/scipy-py38-1.5.4-1/scipy-1.5.4/build/temp.macosx-10.14-x86_64-3.8/ccompiler_opt_cache_clib.py
```
or
```
CCompilerOpt.generate_dispatch_header[2236] : generate CPU dispatch header: (/opt/lib/python3.8/site-packages/numpy/core/src/common/_cpu_dispatch.h)
error: [Errno 13] Permission denied: '/opt/lib/python3.8/site-packages/numpy/core/src/common/_cpu_dispatch.h'
CCompilerOpt._cache_write[796] : write cache to path -> /private/var/tmp/fink.build/scipy-py38-1.5.4-1/scipy-1.5.4/build/temp.macosx-10.14-x86_64-3.8/ccompiler_opt_cache_clib.py
```
– the second case when `numpy.test` was run after initial installation with root permissions, which did create a `src/common/_cpu_dispatch.h` header, but in both cases all of `numpy/core` is of course owned by root and not writeable by a regular user.

I am not sure if this is incorrect usage of `numpy.distutils` by scipy, or a bug in the new `CCompilerOpt`, but since the error shows up with both the stable and RC scipy versions, which both build successfully against numpy 1.19.4, reporting it here first.

Either way `new_ccompiler_opt` should not create files in the numpy install dir under a user who does not have access to it, or should not be called from external package builds.

### NumPy/Python version information:

1.20.0rc2 3.8.6 (default, Oct 15 2020, 01:14:04) 
[Clang 11.0.0 (clang-1100.0.33.17)]

### LINKED ISSUE #18135: numpy wants nonexistent numpy/core/src
See https://bugzilla.redhat.com/show_bug.cgi?id=1913810, prevents building sympy.

### PR DESCRIPTION
**Pull request by seiko2plus**
closes #18092

#### Generate the main dispatcher config header into the build dir

  The new path becomes `build/src.*/numpy/distutils/include/npy_cpu_dispatch_config.h`
  instead of `numpy/core/src/common/_cpu_dispatch.h`.

  The new path allows other projects to re-use the CPU dispatcher
  once we decide to expose the following headers:
    - `numpy/core/src/common/npy_cpu_dispatch.h`
    - `numpy/core/src/common/npy_cpu_features.h`

### CODE REVIEW

File: numpy/core/src/common/npy_cpu_dispatch.h [hunk 1]
```diff
@@ -7,7 +7,7 @@
 #include "npy_cpu_features.h" // NPY_CPU_HAVE
 #include "numpy/utils.h" // NPY_EXPAND, NPY_CAT
 /**
- * Bringing the main configration header '_cpu_dispatch.h'.
+ * Bringing the main configration header 'npy_cpu_dispatch_config.h'.
```
Comment (mattip) [inline numpy/core/src/common/npy_cpu_dispatch.h:10]:
```suggestion
 * Including the main configration header 'npy_cpu_dispatch_config.h'.
```

File: numpy/core/src/common/npy_cpu_dispatch.h [hunk 2]
```diff
@@ -7,7 +7,7 @@
...
  *
  * This header is generated by the distutils module 'ccompiler_opt',
  * and contains all the #definitions and headers of instruction-sets,
```
Comment (mattip) [inline numpy/core/src/common/npy_cpu_dispatch.h:13]:
```suggestion
 * and contains all the #definitions and headers for platform-specific instruction-sets
```

File: numpy/distutils/ccompiler_opt.py [hunk 3]
```diff
@@ -2240,6 +2240,14 @@ def generate_dispatch_header(self, header_path):
         baseline_len = len(baseline_names)
         dispatch_len = len(dispatch_names)
 
+        header_dir = os.path.dirname(header_path)
+        if not os.path.exists(header_dir):
+            self.dist_log(
+                f"dispatch header dir {header_dir} isn't exist, creating it",
```
Comment (mattip) [inline numpy/distutils/ccompiler_opt.py:2246]:
```suggestion
                f"dispatch header dir {header_dir} does not exist, creating it",
```

File: numpy/distutils/ccompiler_opt.py [hunk 4]
```diff
@@ -2504,30 +2512,24 @@ def _generate_config(self, output_dir, dispatch_src, targets, has_baseline=False
             ))
         return False
 
-def new_ccompiler_opt(compiler, **kwargs):
+def new_ccompiler_opt(compiler, dispatch_hpath, **kwargs):
     """
     Create a new instance of 'CCompilerOpt' and generate the dispatch header
-    inside NumPy source dir.
+    which containing all definitions and headers of instruction-sets for
```
Comment (mattip) [inline numpy/distutils/ccompiler_opt.py:2518]:
```suggestion
    which containing the definitions and headers of platform-specific instruction-sets for
```

**Review by mattip — APPROVED** [review]
LGTM. Just a few tweaks in the comments and log message.

**Comment by charris** [comment]
@mattip Is there a pressing need for this in 1.20? I'd prefer to limit 1.20 to bug fixes at this point.

**Comment by mattip** [comment]
Since downstream projects cannot use numpy/distutils without it, it is in my opinion a breaking bug.

File: numpy/distutils/ccompiler_opt.py (hunk 4 above)
Comment (h-vetinari) [reply numpy/distutils/ccompiler_opt.py:2518]:
```suggestion
    which contains the definitions and headers of instruction-sets for
```

File: numpy/core/src/common/npy_cpu_dispatch.h [hunk 5]
```diff
@@ -7,10 +7,10 @@
 #include "npy_cpu_features.h" // NPY_CPU_HAVE
 #include "numpy/utils.h" // NPY_EXPAND, NPY_CAT
 /**
- * Bringing the main configration header '_cpu_dispatch.h'.
+ * Including the main configration header 'npy_cpu_dispatch_config.h'.
```
Comment (dhomeier) [inline numpy/core/src/common/npy_cpu_dispatch.h:10]:
```suggestion
 * Including the main configuration header 'npy_cpu_dispatch_config.h'.
```

**Review by dhomeier — APPROVED** [review]
Thanks for the fix, I applied the patch to 1.20.0rc2 and successfully built scipy against it! Just a minor spelling nitpick.

**Comment by mattip** [comment]
Thanks @seiko2plus

**Pull request merged**